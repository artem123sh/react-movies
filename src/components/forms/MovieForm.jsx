/* eslint-disable camelcase */
import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import {
  Formik, Form, useField, useFormikContext,
} from 'formik';
import Input from '../basic/Input';
import DatePicker from '../basic/DatePicker';
import MultiSelect from '../basic/MultiSelect';
import Label from '../basic/Label';
import { GENRES } from '../../constants';

const StyledLabel = styled(Label)`
  margin-bottom: 0.5rem;
  margin-top: 1.5rem;
  display: block;
`;

const StyledDatePicker = styled(DatePicker)`
  width: calc(100% - 2rem);
  padding: 0 1rem;
`;

const StyledInput = styled(Input)`
  width: calc(100% - 2rem);
  padding: 0 1rem;

`;

const DatePickerContainer = styled.div`
  .react-datepicker-wrapper {
    width: 100%;
  }
`;

const ErrorLabel = styled.div`
  color: red;
  font-size: 1.3rem;
`;

const FormikInput = (props) => {
  const {
    id, label, name, type, placeholder,
  } = props;
  const [field, meta] = useField(props);
  return (
    <>
      <StyledLabel htmlFor={id}>{label}</StyledLabel>
      <StyledInput
        id={id}
        name={name}
        placeholder={placeholder}
        type={type}
        value={field.value ? field.value.toString() : ''}
        onChange={field.onChange}
      />
      {meta.touched && meta.error && <ErrorLabel>{meta.error}</ErrorLabel>}
    </>
  );
};

FormikInput.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
};

export const FormikDatePicker = (props) => {
  const {
    label, id, name, placeholder,
  } = props;
  const { setFieldValue } = useFormikContext();
  const [field, meta] = useField(props);
  return (
    <>
      <StyledLabel htmlFor={id}>{label}</StyledLabel>
      <DatePickerContainer>
        <StyledDatePicker
          id={id}
          name={name}
          placeholder={placeholder}
          value={field.value ? new Date(field.value) : null}
          onChange={(val) => setFieldValue(field.name, val.toISOString().slice(0, 10))}
        />
      </DatePickerContainer>
      {meta.touched && meta.error && <ErrorLabel>{meta.error}</ErrorLabel>}
    </>
  );
};

FormikDatePicker.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
};

export const FormikMultiSelect = (props) => {
  const {
    id, label, placeholder, options,
  } = props;
  const { setFieldValue } = useFormikContext();
  const [field, meta] = useField(props);

  return (
    <>
      <StyledLabel htmlFor={id}>{label}</StyledLabel>
      <MultiSelect
        id={id}
        placeholder={placeholder}
        values={field.value || []}
        options={options}
        onChange={(val) => setFieldValue(field.name, val)}
      />
      {meta.touched && meta.error && <ErrorLabel>{meta.error}</ErrorLabel>}
    </>
  );
};

FormikMultiSelect.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  placeholder: PropTypes.string.isRequired,
};

const MovieForm = ({
  movie, onSubmit, children,
}) => (
  <Formik
    initialValues={movie}
    validate={(values) => {
      const errors = {};
      const required = 'Required Field!';
      const {
        title,
        tagline,
        release_date,
        poster_path,
        overview,
        runtime,
        genres,
      } = values;
      Object.entries({
        title,
        tagline,
        release_date,
        poster_path,
        overview,
        runtime,
        genres,
      })
        .filter(([, value]) => !value)
        .forEach(([key]) => { errors[key] = required; });
      if (!values.genres.length) {
        errors.genres = required;
      }
      if (!Number.parseInt(values.runtime, 10)) {
        errors.runtime = 'Should be a number!';
      }
      const urlRegex = /(http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-/]))?/;
      if (!urlRegex.test(values.poster_path)) {
        errors.poster_path = 'Should be a valid URL!';
      }
      return errors;
    }}
    onSubmit={async (values, { setSubmitting }) => {
      await onSubmit(values);
      setSubmitting(false);
    }}
  >
    <Form>
      <FormikInput id="title" label="Title" name="title" type="text" placeholder="Title" inputClassName={StyledInput} />
      <FormikDatePicker id="release_date" label="Release Date" name="release_date" placeholder="Select Date" />
      <FormikInput id="poster_path" label="Movie Url" name="poster_path" type="text" placeholder="Movie Url here" inputClassName={StyledInput} />
      <FormikMultiSelect id="genres" label="Genre" name="genres" placeholder="Select Genre" options={GENRES} />
      <FormikInput id="overview" label="Overview" name="overview" type="text" placeholder="Overview here" inputClassName={StyledInput} />
      <FormikInput id="runtime" label="Runtime" name="runtime" placeholder="Runtime here" type="number" inputClassName={StyledInput} />
      <FormikInput id="tagline" label="Tagline" name="tagline" type="text" placeholder="Tagline here" inputClassName={StyledInput} />
      {children}
    </Form>
  </Formik>
);

MovieForm.defaultProps = {
  movie: {
    title: '',
    release_date: '',
    poster_path: '',
    genres: [],
    overview: '',
    runtime: 0,
    tagline: '',
  },
};

MovieForm.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string,
    release_date: PropTypes.string,
    poster_path: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    overview: PropTypes.string,
    runtime: PropTypes.number,
    tagline: PropTypes.string,
  }),
  children: PropTypes.node.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

export default MovieForm;
