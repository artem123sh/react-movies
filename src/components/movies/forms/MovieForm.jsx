import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Input from '../../basic/Input';
import DatePicker from '../../basic/DatePicker';
import MultiSelect from '../../basic/MultiSelect';
import Label from '../../basic/Label';

const StyledLabel = styled(Label)`
  margin-bottom: 0.5rem;
`;

const inputStyles = `
  margin: 0.5rem 0 1.5rem 0;
`;

const StyledDatePicker = styled(DatePicker)`
  width: calc(100% - 2rem);
  padding: 0 1rem;
`;

const StyledInput = styled(Input)`
  width: calc(100% - 2rem);
  padding: 0 1rem;
  ${inputStyles}
`;

const StyledMultuSelect = styled(MultiSelect)`
  ${inputStyles}
`;

const DatePickerContainer = styled.div`
  .react-datepicker-wrapper {
    width: 100%;
  }
  ${inputStyles}
`;

const MovieForm = ({ movie, children }) => (
  <form>
    <StyledLabel htmlFor="title">Title</StyledLabel>
    <StyledInput id="title" placeholder="Title" defaultValue={movie.title} />
    <StyledLabel htmlFor="release_date">Release Date</StyledLabel>
    <DatePickerContainer>
      <StyledDatePicker id="release_date" placeholder="Select Date" value={movie.release_date} />
    </DatePickerContainer>
    <StyledLabel htmlFor="poster_path">Movie Url</StyledLabel>
    <StyledInput id="poster_path" placeholder="Movie Url here" defaultValue={movie.poster_path} />
    <StyledLabel htmlFor="genres">Genre</StyledLabel>
    <StyledMultuSelect
      id="genres"
      placeholder="Select Genre"
      values={movie.genres}
      options={['Drama', 'Thriller', 'Crime', 'Mystery', 'Animation']}
    />
    <StyledLabel htmlFor="overview">Overview</StyledLabel>
    <StyledInput id="overview" placeholder="Overview here" defaultValue={movie.overview} />
    <StyledLabel htmlFor="runtime">Runtime</StyledLabel>
    <StyledInput id="runtime" placeholder="Runtime here" defaultValue={movie.runtime ? movie.runtime.toString() : ''} />
    {children}
  </form>
);

MovieForm.defaultProps = {
  movie: {
    title: '',
    release_date: '',
    poster_path: '',
    genres: [],
    overview: '',
    runtime: 0,
  },
};

MovieForm.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string,
    release_date: PropTypes.string,
    poster_path: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    overview: PropTypes.string,
    runtime: PropTypes.number,
  }),
  children: PropTypes.node.isRequired,
};

export default MovieForm;
