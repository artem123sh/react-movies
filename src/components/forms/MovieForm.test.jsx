import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MovieForm from './MovieForm';
import '@testing-library/jest-dom/extend-expect';

const spy = jest.spyOn(console, 'warn');
spy.mockImplementation(() => {});

describe('MovieForm', () => {
  test('renders MovieForm\'s snapshot', () => {
    const { asFragment } = render(<MovieForm movie={{}} onSubmit={() => {}}>test</MovieForm>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('handles title input field correctly', async () => {
    const utils = render(<MovieForm movie={{}} onSubmit={() => {}}>test</MovieForm>);
    const input = utils.getByLabelText('Title');
    input.focus();
    await userEvent.type(input, 'title');
    expect(input.value).toBe('title');
  });
});
