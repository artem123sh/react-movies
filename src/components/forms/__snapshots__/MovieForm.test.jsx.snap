// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm renders MovieForm's snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="title"
    >
      Title
    </label>
    <input
      class="sc-AxjAm sc-fzpans foPLdB"
      id="title"
      placeholder="Title"
      type="text"
      value=""
    />
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="release_date"
    >
      Release Date
    </label>
    <div
      class="sc-fzplWN jjvCdU"
    >
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class="sc-AxirZ vNiBr sc-fzoLsD bCdSFp"
            placeholder="Select Date"
            value=""
          />
        </div>
      </div>
    </div>
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="poster_path"
    >
      Movie Url
    </label>
    <input
      class="sc-AxjAm sc-fzpans foPLdB"
      id="poster_path"
      placeholder="Movie Url here"
      type="text"
      value=""
    />
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="genres"
    >
      Genre
    </label>
    <div
      class="sc-AxhCb jZxptO"
    >
      <div>
        <div
          class="sc-AxgMl fqNPUb"
        >
          <input
            autocomplete="off"
            class="sc-AxjAm sc-AxheI fNdnhK"
            id="genres"
            placeholder="Select Genre"
            readonly=""
            type="text"
            value=""
          />
          <svg
            class="sc-Axmtr fvsATX"
            fill="#F65261"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
      </div>
    </div>
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="overview"
    >
      Overview
    </label>
    <input
      class="sc-AxjAm sc-fzpans foPLdB"
      id="overview"
      placeholder="Overview here"
      type="text"
      value=""
    />
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="runtime"
    >
      Runtime
    </label>
    <input
      class="sc-AxjAm sc-fzpans foPLdB"
      id="runtime"
      placeholder="Runtime here"
      type="number"
      value=""
    />
    <label
      class="sc-AxmLO sc-fzozJi fkdgxq"
      for="tagline"
    >
      Tagline
    </label>
    <input
      class="sc-AxjAm sc-fzpans foPLdB"
      id="tagline"
      placeholder="Tagline here"
      type="text"
      value=""
    />
    test
  </form>
</DocumentFragment>
`;
